src/formatting/cargo_toml.rs

```diff
@@ -0,0 +1,412 @@
+/// put the `name` and `version` keys in that order at the top of that section,
+/// followed by the remaining keys other than `description` in alphabetical order,
+/// followed by the `description` at the end of that section.
+struct SortKey;
+
+/// Put the `[package]` section at the top of the file
+struct SortSection {
+    /// `cargo-edit` uses a `position` field to put tables back in their original order when
+    /// serialising. We should reset this field after sorting.
+    current_position: usize,
+}
+
+/// Put a blank line between the last key-value pair in a section and the header of the next
+/// section.
+///
+/// Do not place a blank line between section headers and the key-value pairs in that section, or
+/// between key-value pairs in a section.
+///
+/// Should be applied after `SortSection`.
+struct BlankLine {
+    trimming: bool,
+}
+
+/// Don't use quotes around any standard key names; use bare keys. Only use quoted
+/// keys for non-standard keys whose names require them, and avoid introducing such
+/// key names when possible.
+///
+/// Put a single space both before and after the = between a key and value.
+/// Do not indent any key names; start all key names at the start of a line.
+struct KeyValue;
+
+/// Use multi-line strings (rather than newline escape sequences) for any string values
+/// that include multiple lines, such as the crate description.
+struct MultiLine;
+
+/// For array values, such as a list of authors, put the entire list on the same line as the key,
+/// if it fits. Otherwise, use block indentation: put a newline after the opening square bracket,
+/// indent each item by one indentation level, put a comma after each item (including the last),
+/// and put the closing square bracket at the start of a line by itself after the last item.
+///
+/// ```toml
+/// authors = [
+///     "A Uthor <a.uthor@example.org>",
+///     "Another Author <author@example.net>",
+/// ]
+///```
+struct WrapArray {
+    max_width: usize,
+}
+
+/// For table values, such as a crate dependency with a path, write the entire
+/// table using curly braces and commas on the same line as the key if it fits. If
+/// the entire table does not fit on the same line as the key, separate it out into
+/// a separate section with key-value pairs:
+///
+/// ```toml
+/// [dependencies]
+/// crate1 = { path = "crate1", version = "1.2.3" }
+///
+/// [dependencies.extremely_long_crate_name_goes_here]
+/// path = "extremely_long_path_name_goes_right_here"
+/// version = "4.5.6"
+/// ```
+struct FormatInlineTable {
+    max_width: usize,
+    /// Must be `InlineTable`
+    long_tables: Vec<(String, String, Item)>,
+    current_section: String,
+}
+
+impl VisitMut for SortKey {
+    fn visit_document_mut(&mut self, doc: &mut Document) {
+        doc.as_table_mut().iter_mut().for_each(|(key, section)| {
+            if key == "package" {
+                let table = section.as_table_mut().expect("package should be a table");
```

Comment:
This definitely shouldn't panic here, since I feel like the logic should definitely propagate an error to the top level so rustfmt can report it properly rather than via a panic message.

It's unfortunate that `VisitMut` doesn't properly allow returning results, but this should encapsulate a proper error somehow.